src/codegen/__future__/definePlugin.ts

```diff
@@ -0,0 +1,211 @@
+   * @param document - the full OAS document
+   */
+  beforeConvertSchema?(
+    schema: SchemaObject | OpenAPIV3.ReferenceObject,
+    fullTree: ts.SourceFile,
+    document: OpenAPIV3.Document,
+  ): void | Promise<void>;
+
+  /**
+   * Run after an OAS schema is converted to an AST node
+   * Can be async
+   * @param schema - the current OAS Schema that was converted
+   * @param typeTree - the AST Type Definition
+   * @param fullTree - the full AST tree
+   * @param document - the full OAS document
+   */
+  afterConvertSchema?(
+    schema: SchemaObject | OpenAPIV3.ReferenceObject,
+    typeTree: ts.TypeNode,
+    fullTree: ts.SourceFile,
+    document: OpenAPIV3.Document,
+  ): void | Promise<void>;
+
+  /**
+   * Runs after the full AST tree has been created, but before the ts file contents have been generated
+   * Can be async
+   * @param fullTree - the full AST tree
+   * @param document - the full OAS document
+   */
+  afterASTGenerate?(
+    fullTree: ts.SourceFile,
+    document: OpenAPIV3.Document,
+  ): void | Promise<void>;
+
+  /**
+   * Runs after the TS file contents have been generated, but before writing to file.
+   * Can be async
+   * @param fileName - name of the file to write to (should this be a full path object?)
+   * @param fileContents - The TS file contents as a string.
+   * @param document - the full OAS document.
+   */
+  beforeWriteToFile?(
+    fileName: string,
+    fileContents: string,
+    document: OpenAPIV3.Document,
+  ): void | Promise<void>;
+
+  /**
+   * Runs after the TS file has been written to. Allows for... cleanup?
+   * @param fileName - name of the file to write to (should this be a full path object?)
+   * @param fileContents - The TS file contents as a string.
+   * @param document - the full OAS document.
+   */
+  afterWriteToFile?(
+    fileName: string,
+    fileContents: string,
+    document: OpenAPIV3.Document,
+  ): void | Promise<void>;
+};
+
+export const IS_OAZAPFTS_PLUGIN = Symbol("OazapftsPlugin");
+
+export type OazapftsCodegenPlugin = OazapftsCodegenPluginDefinition & {
+  [IS_OAZAPFTS_PLUGIN]: boolean;
+};
+
+/**
+ * Assigns a hidden key we can use to make sure the default export is in fact a plugin when we import it and prevent some
+ * potential headaches.
+ * @param plugin
+ * @param value
+ */
+function convertToPlugin(
+  plugin: OazapftsCodegenPluginDefinition,
+  value: boolean,
+): asserts plugin is OazapftsCodegenPlugin {
+  Object.defineProperty(plugin, IS_OAZAPFTS_PLUGIN, {
+    value,
+    enumerable: false,
+    writable: false,
+    configurable: false,
+  });
+}
+
+export function defineOazapftsPlugin(
+  plugin: OazapftsCodegenPluginDefinition,
+): OazapftsCodegenPlugin {
+  // Do some checks to ensure that we have a valid plugin (if there are required params)
+  let isPlugin = true;
+  if (!plugin.name || plugin.name === "") {
+    console.warn("defineOazapftsPlugin: 'name' property is required!");
+    isPlugin = false;
+  }
```

Comment:
Let's not do this for now. Given the name is only needed for logging, we can default to `unknown | anonymous` without hurting anyone badly. 

If we find we need validation let's throw an Error when these fail. Id argue there is no value in a `isPlugin=false` plugin.

Reply:
i've removed the logic for this.